editoast/fga_migrations/src/migration.rs

```diff
@@ -0,0 +1,509 @@
+use std::sync::LazyLock;
+
+use crate::client::CliParser;
+use crate::client::Migration as TargetMigration;
+use fga::client::Client;
+use fga::client::ConnectionSettings;
+use fga::client::InitializationError;
+use fga::client::QueryError;
+use fga::client::RequestFailure;
+use fga::client::TooManyTuples;
+use fga::compile_model;
+use fga::model::Relation;
+use indexmap::IndexMap;
+use indexmap::indexmap;
+use itertools::Either;
+use sha1::Digest;
+use sha1::Sha1;
+use tracing::error;
+use tracing::info;
+use tracing::instrument;
+use url::Url;
+
+const MIGRATIONS_MODEL: &str = include_str!("../migrations_model.fga");
+
+#[cfg(not(test))]
+static OPENFGA_MIGRATIONS: LazyLock<IndexMap<&'static str, &'static str>> = LazyLock::new(|| {
+    // The migrations declaration order matches the order in which they will be applied
+    indexmap! {
+        "initial_model" => include_str!("../migrations/1_initial_model.fga"),
+    }
+});
+
+#[cfg(test)]
+static OPENFGA_MIGRATIONS: LazyLock<IndexMap<&'static str, &'static str>> = LazyLock::new(|| {
+    // The migrations declaration order matches the order in which they will be applied
+    indexmap! {
+        "initial_model" => include_str!("../tests/migrations/1_initial_model.fga"),
+        "intermediate_model" => include_str!("../tests/migrations/2_intermediate_model.fga"),
+        "final_model" => include_str!("../tests/migrations/3_latest_model.fga"),
+    }
+});
+
+#[derive(Debug, Clone)]
+pub struct Settings {
+    pub migration_settings: MigrationSettings,
+    pub telemetry: TelemetrySettings,
+}
+
+#[derive(Debug, Clone)]
+pub struct MigrationSettings {
+    pub connection_settings: ConnectionSettings,
+    pub migration_store: String,
+    pub authorization_store: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct TelemetrySettings {
+    pub endpoint: Url,
+}
+
+impl From<&CliParser> for Settings {
+    fn from(client: &CliParser) -> Self {
+        let connection_settings = ConnectionSettings::new(
+            client.openfga_url.host().unwrap().to_string(),
+            client.openfga_url.port_or_known_default().unwrap(),
+            fga::client::Limits {
+                max_checks_per_batch_check: client.tuple_reads,
+                max_tuples_per_write: client.tuple_writes,
+            },
+        );
+        Settings {
+            migration_settings: MigrationSettings {
+                connection_settings,
+                migration_store: client.migrations_store.clone(),
+                authorization_store: client.authorization_store.clone(),
+            },
+            telemetry: TelemetrySettings {
+                endpoint: url::Url::parse(&client.telemetry_endpoint).expect("Invalid url"),
+            },
+        }
+    }
+}
+
+#[derive(fga::Type, fga::User, derive_more::FromStr, Debug, Clone, PartialEq, Eq, Hash)]
+struct Migration(pub usize);
+
+#[derive(fga::Type, fga::Object, derive_more::FromStr, Debug, Clone, PartialEq, Eq, Hash)]
+struct Model(pub String);
```

Comment:
```suggestion
struct Migration(usize);

#[derive(fga::Type, fga::Object, derive_more::FromStr, Debug, Clone, PartialEq, Eq, Hash)]
struct Model(String);
```

It'd be useful to document what the `usize` and `String` are and why we need *both* the migration number and the hash of its file.

Reply:
Actually I think we could do without the model hash ? It mostly gives an extra guarantee that when we retrieve a migration from its serial number, its associated model hasn't changed since the last time it was called: if we apply "migration 3" on top of itself, we can check that the previous and current runs are based on the same model (or at least that they generated the same hash, which is mostly the same thing).

Changing the model of an existing migration is bad practice though, so we should not need to do that verification between two migration runs.

On the other hand, we chose to track migrations in an openfga store through relation singleton. If we remove the hash of the model used in the model object identifier, we need to either store a random value there or something else that would make sense. The model hashes might be useful when debugging: we might aswell store it compared to storing a random value ?

AFAIK we can only write tuples in an openfga store: writing an orphan object not linked to any relation is not a thing ? 👀 I don't see anything in the http API that would allow to do it and would totally understand if they prevent the users to do it by design.

If we stored the migrations directly in a PostgreSQL table, I would be in favour to only store the migration number of the last applied migration. Switching to Postgres is entirely another matter though: interfacing exclusively with the openfga server in the migrations run means that the `fga_migrations` crate is mostly agnostic of any plumbing on the server side (it just needs to know how to connect to the openfga server and some server settings like its max allowed tuples per write), which is a good thing. Since we are using an openfga store to save the migrations state, we are stuck with the way it stores data. Relations contain at least the identifiers of two objects: if we use them, we have an extra slot to allocate that we can fill with either something meaningful or with a random value.

I think either keeping it this way or removing the hash comparisons (and possibly stop saving the model hashes) during migration runs and only check the migration numbers are fine. What do you think ?

Reply:
model:hash APPLY migration:4

change to

service:editoast APPLY migration:4-<hash>

struct Editoast;

impl Type for Editoast

struct Migration(usize, String);

impl Type

impl FromStr

Reply:
Done :+1: